File: tools/gap_vs_rocm.jl
"""
What this tree gets on one operation against what the vendor stack gets, per
shape, as a table you can re-run.

Every performance note in `plans/` before 2026-09-21 was self-relative — this
tree against itself — and two of the conclusions drawn that way did not survive
a vendor comparison (`plans/2026-09-21-rocm-baseline.md`). This exists so the
gap is a tracked number rather than an occasional discovery.

    julia --project=. tools/gap_vs_rocm.jl            # writes tmp/baseline/ours.json
    tmp/baseline/preload.sh tools/gap_vs_rocm.py      # times the same shapes, prints the table

Two processes because they need different runtimes, and one JSON between them
so the SHAPES are stated once, here. `preload.sh` is why the torch side runs at
all on this box; `plans/2026-09-21-rocm-baseline.md` has that story.

Shapes are the ones the models in this tree actually execute, named after the
op that runs them, so a row that regresses says where to look.
"""

using DNNKernels, Printf, Statistics, Random
const DK = DNNKernels
# Through the package, so this runs from any environment DNNKernels loads in.
const JSON3 = DNNKernels.JSON3
const KA = DK.KernelAbstractions
const M = DK.Mantle

"""Minimum of `n` timed runs, after `warm` untimed ones. Minimum, not mean: a
GPU that shares a machine has a floor and no ceiling."""
function best(backend, f; n = 15, warm = 5)
    for _ in 1:warm; f(); end
    KA.synchronize(backend)
    t = Inf
    for _ in 1:n
        KA.synchronize(backend)
        t0 = time_ns(); f(); KA.synchronize(backend)
        t = min(t, (time_ns() - t0) / 1e6)
    end
    t
end

f16(backend, dims...; scale = 0.3f0) =
    (a = KA.allocate(backend, Float16, dims...);
     copyto!(a, Float16.(randn(Float32, dims...) .* scale)); a)

"""
`(E, Lq, Lk, H, B, calls)` attentions, named for where they run, with how many
times one encode runs each.

Read off the exported graphs, not chosen. `sam2-encoder-global` used to be here
as `E = 64, B = 1` and was BOTH wrong: SAM 2's head dim is 72, and that shape is
3 of the encoder's 48 attention calls. Optimising against it moved the
microbenchmark 23% and the model not at all. The windowed shape below is 32 of
the 48 and is where the time is.
"""
const ATTN = [("qwen-denoiser", 128, 4096, 4118, 32, 1, 60),
              ("qwen-denoiser-square", 128, 4096, 4096, 32, 1, 60),
              ("sam2-windowed-256", 72, 256, 256, 8, 16, 32),
              ("sam2-encoder-global", 72, 4096, 4096, 8, 1, 3),
              ("sam2-windowed-16", 72, 16, 16, 4, 1024, 6)]

"""`(Cin, Cout, H, W)` 3x3 stride-1 convolutions, all of them from the
Qwen-Image 2.1 VAE decoder, which is 82% convolution."""
const CONV = [("vae-288-1024", 288, 288, 1024, 1024),
              ("vae-144-1024", 144, 144, 1024, 1024),
              ("vae-288-144-1024", 288, 144, 1024, 1024),
              ("vae-576-512", 576, 576, 512, 512),
              ("vae-1152-256", 1152, 1152, 256, 256),
              ("vae-1152-128", 1152, 1152, 128, 128)]

"""`(M, N, K)` fp16 products."""
const GEMM = [("square-4096", 4096, 4096, 4096),
              ("denoiser-qkv", 12288, 4224, 4096),
              ("denoiser-gateup", 24576, 4224, 4096)]

function main()
    backend = M.LavaBackend()
    ctx = DK.Ctx(backend)
    caps = DK.caps(backend)
    rows = Any[]

    for (name, E, Lq, Lk, H, B, calls) in ATTN
        q, k, v = f16(backend, E, Lq, H, B), f16(backend, E, Lk, H, B), f16(backend, E, Lk, H, B)
        out = KA.allocate(backend, Float32, E, Lq, H, B); fill!(out, 0f0)
        # **The plan the MODEL would get, which is unclamped first.**
        # `emitsdpa!` asks `flashcm_plan` with no clamp and only retries through
        # `flashcm_padded_plan` when that declines; this file used to force
        # `clamp = Lk % 32 != 0` instead, which is not the same rule and is
        # wrong wherever the chosen `BC` divides `Lk` anyway. `sam2-windowed-16`
        # is exactly that: `Lk = 16` fails `% 32` but the tiling is `(16, 16)`,
        # and the clamped plan is a different tiling on the staged kernel —
        # **1.319 ms against 0.568**, so this row reported a shape SAM 2 does
        # not run, 2.3x slow.
        plan = DK.flashcm_plan(caps, q, k, v, nothing)
        plan isa DK.FlashCMPlan ||
            (plan = DK.flashcm_padded_plan(caps, q, k, v, nothing))
        if plan isa DK.FlashCMPlan
            t = best(backend, () -> DK.sdpaflashcm!(ctx, out, plan, q, k, v, Float32(1/sqrt(E))))
            push!(rows, (kind = "attention", name = name, ms = t,
                         flops = 2.0 * 2 * Lq * Lk * E * H * B,
                         params = Dict("E" => E, "Lq" => Lq, "Lk" => Lk, "H" => H,
                                       "B" => B, "calls" => calls)))
        end
        q = k = v = out = nothing; GC.gc()
    end

    for (name, Cin, Cout, H, W) in CONV
        x = f16(backend, W, H, Cin, 1)
        w = f16(backend, 3, 3, Cin, Cout; scale = 0.05f0)
        out = KA.allocate(backend, Float16, W, H, Cout, 1); fill!(out, Float16(0))
        t = best(backend, () -> DK.convolution!(ctx, out, x, w, nothing, (1, 1), (1, 1), (1, 1), 1); n = 8, warm = 3)
        push!(rows, (kind = "conv3x3", name = name, ms = t,
                     flops = 2.0 * Cin * Cout * 9 * H * W,
                     params = Dict("Cin" => Cin, "Cout" => Cout, "H" => H, "W" => W)))
        x = w = out = nothing; GC.gc()
    end

    for (name, Mm, Nn, Kk) in GEMM
        a, b = f16(backend, Mm, Kk), f16(backend, Kk, Nn)
        _, sk = M.coopmat_gemm_shape(Mm, Nn, Kk)
        c = KA.allocate(backend, Float32, Mm, Nn, max(sk, 1))
        t = best(backend, () -> M.coopmat_gemm!(c, a, b, Mm, Nn, Kk; partials = c, reduce = false); n = 8, warm = 3)
        push!(rows, (kind = "gemm-fp16", name = name, ms = t,
                     flops = 2.0 * Mm * Nn * Kk,
                     params = Dict("M" => Mm, "N" => Nn, "K" => Kk)))
        a = b = c = nothing; GC.gc()
    end

    for (name, T, nbytes) in (("copy-fp16-512MiB", Float16, 512 << 20),
                              ("copy-fp32-512MiB", Float32, 512 << 20))
        n = nbytes ÷ sizeof(T)
        src = KA.allocate(backend, T, n); dst = KA.allocate(backend, T, n)
        t = best(backend, () -> copyto!(dst, src))
        push!(rows, (kind = "copy", name = name, ms = t, flops = 0.0,
                     params = Dict("bytes" => nbytes, "eltype" => string(T))))
        src = dst = nothing; GC.gc()
    end

    out = joinpath(@__DIR__, "..", "tmp", "baseline", "ours.json")
    mkpath(dirname(out))
    open(out, "w") do io
        JSON3.pretty(io, (rows = rows,))
    end
    @printf("%-10s %-22s %10s %10s\n", "kind", "name", "ms", "TFLOP/s")
    for r in rows
        @printf("%-10s %-22s %10.2f %10s\n", r.kind, r.name, r.ms,
                r.flops == 0 ? "-" : @sprintf("%.2f", r.flops / (r.ms * 1e-3) / 1e12))
    end
    println("\nwrote ", out)
end

main()

File: tools/gap_vs_rocm.py
"""
The vendor half of `gap_vs_rocm.jl`: times the shapes that file states, on the
same GPU, and prints the ratio.

    julia --project=. tools/gap_vs_rocm.jl
    tmp/baseline/preload.sh tools/gap_vs_rocm.py

`preload.sh` is not optional here. The PyTorch ROCm wheels bundle a ROCm
runtime that segfaults at the first kernel launch on gfx1151 while the system
one is fine; see `plans/2026-09-21-rocm-baseline.md`.

Shapes come from the JSON so there is one place they are written down. A row
this file cannot run is reported as such rather than skipped, because a missing
row reads like a passing one.
"""
import json, os, sys, time
import torch
import torch.nn.functional as F

HERE = os.path.dirname(os.path.abspath(__file__))
OURS = os.path.join(HERE, "..", "tmp", "baseline", "ours.json")
dev = "cuda"


def best(fn, n=15, warm=5):
    for _ in range(warm):
        fn()
    torch.cuda.synchronize()
    t = float("inf")
    for _ in range(n):
        torch.cuda.synchronize()
        t0 = time.perf_counter()
        fn()
        torch.cuda.synchronize()
        t = min(t, (time.perf_counter() - t0) * 1e3)
    return t


def run(row):
    p, kind = row["params"], row["kind"]
    if kind == "attention":
        E, Lq, Lk, H = p["E"], p["Lq"], p["Lk"], p["H"]
        q = torch.randn(1, H, Lq, E, device=dev, dtype=torch.float16) * 0.3
        k = torch.randn(1, H, Lk, E, device=dev, dtype=torch.float16) * 0.3
        v = torch.randn(1, H, Lk, E, device=dev, dtype=torch.float16) * 0.3
        return best(lambda: F.scaled_dot_product_attention(q, k, v))
    if kind == "conv3x3":
        ci, co, h, w = p["Cin"], p["Cout"], p["H"], p["W"]
        x = torch.randn(1, ci, h, w, device=dev, dtype=torch.float16) * 0.3
        kk = torch.randn(co, ci, 3, 3, device=dev, dtype=torch.float16) * 0.05
        return best(lambda: F.conv2d(x, kk, padding=1), n=8, warm=3)
    if kind == "gemm-fp16":
        m, n, kd = p["M"], p["N"], p["K"]
        a = torch.randn(m, kd, device=dev, dtype=torch.float16) * 0.3
        b = torch.randn(kd, n, device=dev, dtype=torch.float16) * 0.3
        return best(lambda: torch.mm(a, b), n=8, warm=3)
    if kind == "copy":
        dt = torch.float16 if p["eltype"] == "Float16" else torch.float32
        n = p["bytes"] // dt.itemsize
        a = torch.empty(n, device=dev, dtype=dt)
        b = torch.empty(n, device=dev, dtype=dt)
        return best(lambda: b.copy_(a))
    return None


def main():
    if not os.path.exists(OURS):
        sys.exit(f"no {OURS}; run tools/gap_vs_rocm.jl first")
    rows = json.load(open(OURS))["rows"]
    torch.backends.cudnn.benchmark = True
    print(f"{'kind':10} {'name':22} {'ours ms':>9} {'torch ms':>9} {'gap':>7} {'ours TF':>8} {'torch TF':>9}")
    worst = []
    for r in rows:
        t = run(r)
        if t is None:
            print(f"{r['kind']:10} {r['name']:22} {'':>9} {'unsupported here':>9}")
            continue
        gap = r["ms"] / t
        fl = r["flops"]
        otf = f"{fl / (r['ms'] * 1e-3) / 1e12:.2f}" if fl else "-"
        ttf = f"{fl / (t * 1e-3) / 1e12:.2f}" if fl else "-"
        print(f"{r['kind']:10} {r['name']:22} {r['ms']:9.2f} {t:9.2f} {gap:6.2f}x {otf:>8} {ttf:>9}")
        worst.append((gap, r["kind"], r["name"]))
        torch.cuda.empty_cache()
    worst.sort(reverse=True)
    print("\nwidest gaps:")
    for g, k, n in worst[:5]:
        print(f"  {g:5.2f}x  {k} {n}")


main()
